File: get_started/neural_networks_in_matlab/rbergomi_extended_example.m
%% Clear workspace and load code:
clear;
project_folder = fileparts(fileparts(fileparts(matlab.desktop.editor.getActiveFilename)));
addpath(genpath(project_folder));

%% Load the extended rough Bergomi neural network(s):
[model, Txi] = LoadrBergomiExtendedNeuralNetwork(project_folder);

% Inspect the model object:
model

% The vector below shows the grid points of the piecewise flat forward variance curve:
Txi

%% Plot the contract grid:
figure;
scatter(model.T,model.k,'.');
xlabel('Expiration');ylabel('Log-moneyness');
title('Contract grid');

%% Plot an example:
% Define the contracts:
k_obs = model.k;
T_obs = model.T;
cartProd = false;

% Choose the model parameters:
alpha = -0.45;
beta = -0.35;
eta = 2.1;
rho = -0.90;
xi = 0.30.^2;
par = [eta;rho;alpha;beta;repmat(xi,27,1)];

% Evaluate the neural network:
tic;
[iv, k_obs, T_obs] = model.Eval(par,k_obs,T_obs,cartProd);
toc;

% Compute true prices:
model_true = rBergomiExtClass('alpha',alpha,'beta',beta,'eta',eta,'rho',rho,'xi',xi);
model_true.pricerSettings.N = 25000; % number of paths to simulate
[iv_true,~,~,~,se] = model_true.GetPrices(k_obs,T_obs,false,'priceType','implied_volatility',...
                                              'standardErrors',true);

                                          
% Plot:
uniqT = unique(T_obs);
idxPlot = [9,22,39,64];
figure;
for i=1:size(idxPlot,2)
    subplot(2,2,i);
    idx = T_obs == uniqT(idxPlot(i));
    plot(k_obs(idx),iv_true(idx),'-','Color','blue','DisplayName','Actual','LineWidth',2);hold on;
    plot(k_obs(idx),iv(idx),'--','Color','red','DisplayName','Neural network','LineWidth',2);
    plot(k_obs(idx),iv_true(idx)-se(idx)*1.96,'--','Color','blue','LineWidth',0.7);hold on;
    plot(k_obs(idx),iv_true(idx)+se(idx)*1.96,'--','Color','blue','LineWidth',0.7);hold on;
    xlabel('Log-moneyness');
    ylabel('Implied volatility');
    title(['T = ',num2str(uniqT(idxPlot(i)))]);
    legend('Monte Carlo','Neural network','95 % Confidence interval (Monte Carlo)')
end


%% Load and filter example contracts:
% Load example contracts:
tmp = importdata([project_folder,'\get_started\example_contracts.txt']);
k_orig = tmp(:,1);
T_orig = tmp(:,2);

figure;
plot(T_orig,k_orig,'.');
title('Observed contracts'); 
xlabel('Expiry');
ylabel('Log-moneyness');
hold on;

% Filter contracts:
[k_obs,T_obs] = model.FilterContracts(k_orig,T_orig,false);
plot(T_obs,k_obs,'.');
legend('Original observed contracts','Those within the supported domain.');

%% How to speed up (repeated) evaluation:
% When repeatedly computing prices over the same contracts (although over
% possibly different model parameters) it is possible to perform a number
% of pre-computations to speed things up. This is very useful when
% calibrating. We illustrate this below.

nRuns = 1000;

% Without pre-computations:
tic
for i=1:nRuns
    iv = model.Eval(par,k_obs,T_obs,false,[]);
end
toc;

% With pre-computations:
% Important remark: Contracts must be sorted appropriately 
% first, see also the description of the PerformPrecomputations 
% method of the NeuralNetworkPricingModel class:
tmp = sortrows([T_obs,k_obs]);
T_obs = tmp(:,1);
k_obs = tmp(:,2);
preCompInfo = model.PerformPrecomputations(k_obs,T_obs);
tic
for i=1:nRuns
    iv = model.Eval(par,k_obs,T_obs,false,preCompInfo);
end
toc;

% It is also possible to skip a number of validation checks and 
% computations in the evaluation function if you can ensure certain 
% assumptions on the inputs are met, e.g. contracts should be sorted 
% etc. One should use this option with care - please read the 
% description for the Eval method of the NeuralNetworkPricingModel 
% class for the exact assumptions on the inputs.

%Example:
skipChecks = true;
tmp = sortrows([T_obs,k_obs]);
T_obs = tmp(:,1);
k_obs = tmp(:,2);

% With pre-computations and without checks:
tic
for i=1:nRuns
    iv = model.Eval(par,k_obs,T_obs,false,preCompInfo,skipChecks);
end
toc;

%% Calibration example:
% Set parameters:
alpha = -0.45;
beta = -0.25;
eta = 2.1;
rho = -0.90;
xi = 0.15.^2;
par_true = [eta;rho;alpha;beta;repmat(xi,27,1)];

% Filter contracts to the neural network domain:
[k_obs,T_obs] = model.FilterContracts(k_orig,T_orig,false);

% Sort contracts:
tmp = sortrows([T_obs,k_obs]);
T_obs = tmp(:,1);
k_obs = tmp(:,2);

% Perform pre-computations (on sorted contracts):
preCompInfo = model.PerformPrecomputations(k_obs,T_obs);

% Generate synthetic 'observed' prices:
iv_obs = model.Eval(par_true,k_obs,T_obs,false,preCompInfo);

% Set optimizer settings:
options = optimoptions('lsqnonlin','Algorithm','trust-region-reflective',...
'MaxFunctionEvaluations',10^4,'MaxIterations',10^4,'FunctionTolerance',10^(-12),...
'StepTolerance',10^(-4),'FiniteDifferenceStepSize',10^(-4),'Display','off');

% Merge sections of the forward variance curve:
uniqT = unique(T_obs);
grpPart = MergeForwardVarianceParameters(Txi,uniqT).';
idxConvParToNN = [1,2,3,4,grpPart + 4];

% Set the initial guess:
x0 = [2.5,-0.8,0,0,ones(1,max(grpPart))*0.3.^2].';

% Define the error function ('lsqnonlin' automatically squares and sums the result):
err = @(x)(iv_obs - model.Eval(x(idxConvParToNN),k_obs,T_obs,false,preCompInfo,true));

% Set parameter bounds:
lb = accumarray(idxConvParToNN',model.lb,[],@max);
ub = accumarray(idxConvParToNN',model.ub,[],@min);

% Run optimizer:
tic;
[xOpt,resnorm,residual,exitflag,output] = lsqnonlin(err,x0,lb,ub,options);
toc;

par_opt = xOpt(idxConvParToNN);

% Compare true and calibrated parameters:
[par_true, par_opt]

%% Illustrate fit:
% Pick an expiration:
T_plot = uniqT(4);

% Plot:
idx = T_obs == T_plot;
figure;
plot(k_obs(idx),iv_obs(idx),'x','DisplayName','Observed');
hold on;
dk = (max(k_obs(idx)) - min(k_obs(idx)))/100;
k_eval = (min(k_obs(idx)):dk:max(k_obs(idx)))';
iv_fit = model.Eval(par_opt,k_eval,T_plot);
plot(k_obs(idx),iv_obs(idx),'-','DisplayName','Calibrated model');
xlabel('Log-moneyness');ylabel('Implied volatility');
title(['Calibration fit (T = ', num2str(T_plot) ')']);
legend();

File: get_started/neural_networks_in_r/rbergomi_extended_example.R
# Description: Shows how to use the extended rough Bergomi neural network implementation.

# Remarks: 
# o Although still fast, the code has not been optimised (the Matlab version has).
# o The underlying models are not implemented in R. You must therefore 
#	  consult the Matlab code to check the neural networks against the 
#	  actual pricing models.

# Load libraries:
#install.packages("jsonlite")
#install.packages("ggplot2")
#install.packages("gridExtra")
library(jsonlite)
library(ggplot2)
library(gridExtra)

# Set paths to source code, weights and more:
# Remark: Works in Rstudio, otherwise you can set the folders manually.
project_folder <- dirname(dirname(dirname(rstudioapi::getSourceEditorContext()$path)))
code_folder <- gsub("/","//",paste(project_folder,"/code/r_code",sep=""))
contract_grid_folder <- gsub("/","//",paste(project_folder,"/code/neural_networks/data",sep=""))
weights_folder <- gsub("/","//",paste(project_folder,"/code/neural_networks/data/neural_network_weights/rbergomi_extended",sep=""))
example_contracts_folder <- gsub("/","//",paste(project_folder,"/get_started",sep=""))
source(gsub("/","//",paste(code_folder,"/NeuralNetworkPricing.r",sep="")))

# Load model:
model <- LoadModel(contract_grid_folder,weights_folder,"rbergomi_extended")

# Plot contract grid:
data_tmp <- data.frame(model$T,model$k)
ggplot(data_tmp, aes(x=model.T, y=model.k)) + geom_point(color="darkblue") + 
	   ggtitle("Contract grid") + xlab("Expiries") + ylab("Log-moneyness") +
	   theme(plot.title = element_text(hjust = 0.5))

# Set parameters:
par <- c(2.1,-0.90,-0.4,0.4,rep(0.15^2,length(model$Txi)))

# Remark: The order is (alpha,beta,eta,rho,xi1,xi2,...,xi27) where the forward variance curve 
# is assumed piecewise flat with gridpoints:
model$Txi

# Evaluate network in grid:
iv <- GetImpliedVolatility(model,par,model$k,model$T)

# Plot a few expiries:
uniqT <- unique(model$T)
idx_plot <- c(1,10,20,50)
data <- data.frame(model$k,model$T,iv)
p <- list()
for (i in 1:length(idx_plot)){
  p[[i]] <- ggplot(data[uniqT[idx_plot[i]] == data$model.T,], aes(x=model.k, y=iv)) +
            geom_line(color="darkblue") + geom_point(color="darkblue") + 
            ggtitle(paste("T = ",uniqT[idx_plot[i]],sep="")) +
            xlab("Log-moneyness") + ylab("Implied volatility") +
            theme(plot.title = element_text(hjust = 0.5))
}
grid.arrange(p[[1]],p[[2]],p[[3]],p[[4]],ncol=2)

# Load example contracts:
tmp <- read.delim(paste(example_contracts_folder,"//example_contracts.txt",sep=""), header = FALSE)
k_orig <- tmp$V1
T_orig <- tmp$V2

# Remove observed contracts outside the neural network domain:
idxKeep <- AreContractsInDomain(model,k_orig,T_orig)
k_obs <- k_orig[idxKeep]
T_obs <- T_orig[idxKeep]

# Store in a data.frame:
data1 <- data.frame(k_orig,T_orig,rep(FALSE,length(k_orig),1))
names(data1) <- c("k","T","Filtered")
data2 <- data.frame(k_obs,T_obs,rep(TRUE,length(k_obs),1))
names(data2) <- c("k","T","Filtered")
data <- rbind(data1,data2)

# Plot observed contracts (before and after filtering):
ggplot(data, aes(x = T, y = k)) + geom_point(aes(color = Filtered)) + 
       ggtitle("Contracts") + xlab("Expiries") + ylab("Log-moneyness") +
       theme(plot.title = element_text(hjust = 0.5)) + ylim(c(-2,0.5))

# The number of contracts is:
length(k_obs)

# The number of (unique) expiries is:
uniqT <- unique(T_obs)
length(uniqT)

# Generate synthetic prices:
parTrue <- c(2.1,-0.90,-0.4,0.4,rep(0.15^2,length(model$Txi)))
iv_obs <- GetImpliedVolatility(model,parTrue,k_obs,T_obs)
iv_obs

# Calibrate the model to see if we can recover the parameters:
# Remark: We simplify by forcing the forward variance curve flat.
# This also avoids an overparameterised model.
idxConvPar <- c(1,2,3,4,rep(5,length(model$Txi)))
err_fun <- function(par){sum((GetImpliedVolatility(model,par[idxConvPar],
                                                   k_obs,T_obs)-iv_obs)^2)}
par0 <- c(1.5,-0.60,-0.2,0.45,0.3^2)
calib <- optim(par0,err_fun,method="L-BFGS-B",
               lower=model$lb[1:length(par0)],
               upper=model$ub[1:length(par0)])

# Compare calibrated parameters to the true ones:
calib$par[idxConvPar]
parTrue

# Plot fit (blue = observed, red = fit):
idx_plot <- c(1,7,10,15)
T_plot <- uniqT[idx_plot]
iv_fit <- GetImpliedVolatility(model,calib$par[idxConvPar],k_obs,T_obs)
data1 <- setNames(data.frame(k_obs,T_obs,iv_obs,rep(TRUE,length(k_obs),1)),c("k_obs","T_obs","iv","obs"))
data2 <- setNames(data.frame(k_obs,T_obs,iv_fit,rep(FALSE,length(k_obs),1)),c("k_obs","T_obs","iv","obs"))
data <- rbind(data1,data2)
p <- list()
for (i in 1:length(idx_plot)){
  dataTmp <- data[T_plot[i] == data$T_obs,]
  p[[i]] <- ggplot(dataTmp, aes(x=k_obs, y=iv)) + geom_point(data=dataTmp[dataTmp$obs==TRUE,],color="blue") + 
            geom_line(data=dataTmp[dataTmp$obs==FALSE,],color="red")  + ggtitle(paste("T = ",round(uniqT[idx_plot[i]],4),sep="")) +
            xlab("Log-moneyness") + ylab("Implied volatility") + theme(plot.title = element_text(hjust = 0.5)) +
            scale_colour_manual("",values=c("blue","red"))
}
grid.arrange(p[[1]],p[[2]],p[[3]],p[[4]],ncol=2)

